tests/nlforces/test_general_poly_stiffness.py

```diff
@@ -0,0 +1,406 @@
+        grad_failed = vutils.check_grad(fun, U, verbose=False, 
+                                        rtol=self.rtol_grad)
+
+        self.assertFalse(grad_failed, 'Incorrect displacement gradient.')        
+        
+        # Numerically Verify Frequency Gradient
+        fun = lambda w: nl_force.aft(U, w[0], h)[0::2]
+        grad_failed = vutils.check_grad(fun, np.array([w]), verbose=False,
+                                        rtol=self.rtol_grad)
+    
+        self.assertFalse(grad_failed, 'Incorrect displacement frequency.')   
+               
+        ######################
+        # Test without zeroth harmonic + skipping 4th harmonic
+        h = np.array([1, 2, 3, 5, 6, 7]) # Automate Checking with this
+        
+        Nhc = 2*(h !=0).sum() + (h==0).sum() # Number of Harmonic Components
+        Nd = Q.shape[1]
+        
+        U = np.random.rand(Nd*Nhc, 1)
+        
+        fun = lambda U: nl_force.aft(U, w, h)[0:2]
+        grad_failed = vutils.check_grad(fun, U, verbose=False, 
+                                        rtol=self.rtol_grad)
+  
+        self.assertFalse(grad_failed, 'Incorrect displacement gradient.')   
+               
+        # Numerically Verify Frequency Gradient
+        fun = lambda w: nl_force.aft(U, w[0], h)[0::2]
+        grad_failed = vutils.check_grad(fun, np.array([w]), verbose=False, 
+                                        rtol=self.rtol_grad)
+        
+        self.assertFalse(grad_failed, 'Incorrect frequency gradient.')   
+        
+    def test_zeroU_input(self):    
+        """
+        Test derivatives when the input displacement is zero. Earlier gradiaent
+        calculation in matlab was not calculating correct gradient at u=0   
+
+        Returns
+        -------
+        None.
+
+        """
+        #######################
+        # Test System
+        
+        # Simple Mapping to spring displacements
+        Q = np.array([[1, 0,0], \
+                      [0, 1,0], \
+                      [0, 0, 1 ]])
+        
+        # Weighted / integrated mapping back for testing purposes
+        T = np.array([[1, 0,0], \
+                      [0, 1,0], \
+                      [0, 0, 1 ]])
+                 
+        qq = np.array([[2, 0, 0], \
+                       [0, 2, 0], \
+                       [3, 0, 0], \
+                       [0, 0, 3],  \
+                       [1, 1, 0], \
+                       [0, 1, 1],  \
+                       [2, 1, 0], \
+                       [1, 1,1]])
+            
+        Emat = np.ones((Q.shape[1],qq.shape[0]))
+            
+        # h = np.array([0, 1, 2, 3]) # Manual Checking expansion / debugging
+        h = np.array([0, 1, 2, 3, 4, 5, 6, 7]) # Automate Checking with this
+        Nhc = 2*(h !=0).sum() + (h==0).sum() # Number of Harmonic Components
+        
+        Nd = Q.shape[1]
+        
+        U = np.zeros((Nd*Nhc, 1))
+        
+        # First DOF, Cosine Term, Fundamental
+        U[Nd+0, 0] = 4
```

Comment:
The comment makes it sound like this test is for at U=0, but then here you are setting components of U != 0?

Reply:
The function tests gradients when some dofs are zero. It has be also tested for all dofs=0. The error is 1e-10 (Greater than the tolerance value set 1e-12)